hybrid/decomposers.py

```diff
@@ -44,6 +45,103 @@ class IdentityDecomposer(traits.ProblemDecomposer, traits.SISO, Runnable):
     def next(self, state, **runopts):
         return state.updated(subproblem=state.problem)
 
+class ComponentDecomposer(traits.ProblemDecomposer, traits.SISO, Runnable):
+    """Selects a subproblem of variables that make up a connected component.
+
+    Args:
+        rolling (bool, optional, default=True):
+            If True, successive calls for the same problem (with possibly
+            different samples) produce subproblems on different components,
+            selected by rolling down the list of all components.
+
+        key (callable, optional, default=None):
+            Extracts a comparison key from each component of the problem. The 
+            comparison keys determine which component will be used to build the 
+            subproblem.
+
+        reverse (bool, optional, default=None):
+            Only applies when `key` is specified. If False, the components' comparison 
+            keys will be sorted in increasing order. If True or unspecified, they will
+            be sorted in decreasing order.
+
+        silent_rewind (bool, optional, default=True):
+            If False, raises :exc:`EndOfStream` when resetting/rewinding the
+            subproblem generator once all components have been used. 
+    
+    See :ref:`decomposers-examples`.
+
+    """
+
+    def __init__(self, rolling=True, silent_rewind=True, key=None, reverse=None, **runopts):
+        super(ComponentDecomposer, self).__init__(**runopts)
+
+        self.rolling = rolling
+        self.silent_rewind = silent_rewind
+        self.key = key
+        self.reverse = reverse
+
+        self._rolling_bqm = None
+        self._rolling_components = []
+        self._component_index = 0
+
+    def __repr__(self):
+        return ("{self}(rolling={self.rolling!r}, "
+                "silent_rewind={self.silent_rewind!r}, "
+                "key={self.key!r}, "
+                "reverse={self.reverse!r})").format(self=self)
+
+    def _get_rolling_components(self, bqm):
+        components = list(connected_components(bqm))
+
+        if self.key is None:
+            return components
+        else:
+            if self.reverse is None:
+                self.reverse = True # Decreasing order by default
+
+            return sorted(components, key=self.key, reverse=self.reverse)
```

Comment:
By always returning an iterable, you can simplify traversal (no need to keep the index variable).
```suggestion
            return iter(sorted(components, key=self.key, reverse=self.reverse))
```